src/surface.rs

```diff
@@ -165,6 +169,49 @@ impl Surface {
         unsafe { ffi::cairo_surface_get_fallback_resolution(self.to_raw_none(), &mut x_pixels_per_inch, &mut y_pixels_per_inch); }
         (x_pixels_per_inch, y_pixels_per_inch)
     }
+
+    pub fn get_extents(&self) -> Option<Rectangle> {
+        unsafe {
+            let rectangle: Rectangle = ::std::mem::zeroed();
+            if ffi::cairo_recording_surface_get_extents(self.to_raw_none(),
+                                                        rectangle.to_raw_none()).as_bool() {
+                Some(rectangle)
+            } else {
+                None
+            }
+        }
+    }
+
+    pub fn create_similar_image(&self, format: Format, width: i32, height: i32) -> Option<Surface> {
+        unsafe {
+            let p = ffi::cairo_surface_create_similar_image(self.to_raw_none(), format.into(), width, height);
+            if p.is_null() {
+                None
+            } else {
+                Some(Self::from_raw_full(p))
+            }
+        }
+    }
+
+    pub fn map_to_image(&self, extents: Option<RectangleInt>) -> Option<Surface> {
+        unsafe {
+            let p = match extents {
+                Some(ref e) => ffi::cairo_surface_map_to_image(self.to_raw_none(), e.to_raw_none()),
+                None => ffi::cairo_surface_map_to_image(self.to_raw_none(), 0 as *const _),
+            };
+            if p.is_null() {
+                None
+            } else {
+                Some(Self::from_raw_full(p))
+            }
+        }
+    }
+
+    pub fn unmap_image(&self, image: &Surface) {
+        unsafe {
+            ffi::cairo_surface_unmap_image(self.to_raw_none(), image.to_raw_none())
```

Comment:
```
The content of the image will be uploaded to the target surface. Afterwards, the image is destroyed.

Using an image surface which wasn't returned by cairo_surface_map_to_image() results in undefined behavior.
```

This needs to check that the surface is of the correct type (needs a surface wrapper type returned by map I guess), and it takes ownership of the image surface.

Reply:
I directly used `ImageSurface` type.

Reply:
That's not sufficient. You can only unmap an image surface that you previously got via map here (for the same `self`).

And you need to use `image.to_raw_full()` or something.

Reply:
Also unclear if that's indeed going to be an image surface, always.

Reply:
> Also unclear if that's indeed going to be an image surface, always.

Let's assume it'll be for now.

Reply:
Then you have to make this API unsafe.

Reply:
... if you don't ensure that it's a surface we got from `map`, I mean.

Also still doing `image.to_raw_none()`. This needs to be `full`.

Reply:
I think what would make sense here is to have `map` return a `struct MappedImageSurface(original_surface, image_surface)`, have it implement `Deref` for the image surface. And on `Drop` it would then call `unmap` and there would be no explicit `unmap` anywhere.

Reply:
Nice idea! I'll implement it this way.